### PR TITLE
Tests

### PR DESCRIPTION
**Pull request by jrconlin**
Final post test can't be completed due to an error with `with async`.
I'm not sure why, but this fails because of a missing attribute. The
object returned contains the attribute, but the function still throws an
attribute error.

This pull will be smoothed out after reviews have been approved. It's ugly right now because of a required early publish.

### CODE REVIEW

**Comment by pjenvey** [comment]
The async with's issue a known problem w/ mock:

http://bugs.python.org/issue26467

async with does a [special method lookup](https://docs.python.org/2/reference/datamodel.html#new-style-special-lookup) of `__aexit__`:  a `type(session).__aexit__` rather than `session.__aexit__`, hence the AttributeError

**Comment by jrconlin** [comment]
Ah, cool. Yeah, that'd definitely do it.

File: push_test/tests/test_pushclient.py [hunk 1]
```diff
@@ -0,0 +1,370 @@
...
+        call_args = json.loads(self.mocks['send'].call_args[0][0])
+        eq_(call_args, dict(messageType="hello", use_webpush=1))
+        eq_(self.mocks['close'].called, False)
+        eq_(self.mocks['close_connection'].called, False)
+
+    @patch('websockets.connect')
+    def test_run_bad_recv(self, m_connect):
+        async def go():
+            try:
+                # Can't call .assert_raises() using an await
```
Comment (pjenvey) [inline push_test/tests/test_pushclient.py]: await within assert_raises as a context manager should work

File: README.rst [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+``args``
+- `channelID` uuid of the channel to create (None if new)
+
+- `key` - VAPID public key base64 string (optional)
+
+*push* - push to an endpoint
+
+``args``
+
+- `pushEndpoint` - the push endpoint (defaults to registered)
```
Comment (pjenvey) [inline README.rst]: this doesn't look supported yet

File: push_test/pushclient.py [hunk 3]
```diff
@@ -0,0 +1,334 @@
...
+        """Receiver handler for websocket messages
+
+        """
+        try:
+            while self.connection:
+                message = await self.connection.recv()
+                await self.process(json.loads(message))
+        except websockets.exceptions.ConnectionClosed:
+            log_msg(out=self.output, status="Websocket Connection closed")
+            pass
```
Comment (pjenvey) [inline push_test/pushclient.py]: kill

File: push_test/tests/test_pushclient.py [hunk 4]
```diff
@@ -0,0 +1,370 @@
...
+        self.client = PushClient(loop=self.loop, args=TrialSettings())
+        self.client.output = self.output
+        self.tasks = [
+            ["hello", {}],
+        ]
+
+    def scan_log(self, phrase):
+        self.output.seek(0)
+        lines = self.output.readlines()
+        for line in lines:
```
Comment (pjenvey) [inline push_test/tests/test_pushclient.py]: return any(phrase in line for line in lines) could kill that pragma nocover!

**Review by pjenvey — COMMENTED** [review]
could use a flake8 run

I'm surprised you went w/ the JSON commands: mostly because this also looks like the basis of an aplt replacement framework that could be utilized under a molotov script =]

**Comment by jrconlin** [comment]
First: see PR #2, apparently github gets SUPER FUSSY about some merges. Going to have to address that in the future.

Second: the JSON thing is mostly shorthand for me. I could use pretty much anything there that allowed me to specify a command and a dict of arguments. What can I say? I think in Javascript sometimes. For what it's worth, I fully expect and desire that there are more capable tools out there. This is by no means Turing complete and is mostly just a means for QA to automate some tests that are currently BASH scripts.

**Pull request closed without merging**